File: src/main/java/net/adventuregame/mobs/Mob.java
package net.adventuregame.mobs;

import com.chidozie.core.terrains.Terrain;
import net.adventuregame.entities.Entity;
import net.adventuregame.game.AdventureMain;
import net.adventuregame.game.GameState;
import net.adventuregame.models.TexturedModel;
import org.joml.Vector3f;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.List;

public abstract class Mob extends Entity {
    private static final Logger log = LoggerFactory.getLogger(Mob.class);
    protected float maxHealth;
    protected boolean isHostile;
    protected static Vector3f position;
    protected float speed;
    protected String name;
    protected float health = maxHealth;

    public Mob(TexturedModel model, int index, Vector3f position, Vector3f rotation, float scale, float maxHealth, boolean isHostile, String name, float speed) {
        super(model, index, position, rotation.x, rotation.y, rotation.z, scale);
        this.maxHealth = maxHealth;
        this.isHostile = isHostile;
        this.name = name;
        this.position = position;
        this.speed = speed;
    }

    public void Move(Vector3f direction) {
        System.out.print("hh");
    }

    public abstract void move(Terrain terrain);


    public abstract void TakeDamage(float damage, Mob Attacker);

    public abstract void Die();

    public abstract void attack(Mob target);

    public String getName() {
        return name;
    }

    protected Mob findClosestEnemy() {
        Mob closest = null;
        float closestDistance = Float.MAX_VALUE;

        for (Mob mob : getNearbyMobs(10.0f)) { // Implement this
            float distance = position.distance(mob.getPosition());
            if (distance < 2.0f && distance < closestDistance) {
                closest = mob;
                closestDistance = distance;
                log.info("Closes Mob FOUND! It's name is: {}", mob.getName());
            }
        }
        return closest;
    }

    private List<Mob> getNearbyMobs(float detectionRadius) {
        List<Mob> nearbyMobs = new ArrayList<>();

        for (Entity entity : GameState.getEntities()) { // Assuming getEntities() returns all entities
            if (entity instanceof Mob mob && mob != this) { // Check if it's a Mob and not the player itself
                if (position.distance(mob.getPosition()) <= detectionRadius) {
                    nearbyMobs.add(mob);
                }
            }
        }
        return nearbyMobs;
    }
}

File: src/main/kotlin/net/adventuregame/mobs/Mob.kt
package net.adventuregame.mobs

import com.adv.core.terrains.Terrain
import net.adventuregame.entity.Entity
import net.adventuregame.game.GameState.Companion.getEntities
import net.adventuregame.models.TexturedModel
import net.adventuregame.scripting.api.LuaEntityAPI
import org.joml.Vector3f
import org.luaj.vm2.LuaValue
import org.luaj.vm2.lib.jse.CoerceJavaToLua
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.util.UUID

abstract class Mob(
    val id: UUID = UUID.randomUUID(),
    index: Int,
    override var position: Vector3f,
    rotation: Vector3f,
    scale: Float,
    protected var maxHealth: Float,
    var isHostile: Boolean,
    var name: String = "Mob",
    protected var speed: Float
) : Entity(
    index,
    position, rotation.x, rotation.y, rotation.z, scale
) {
    abstract override fun GetModels(): List<TexturedModel>

    var health: Float = maxHealth
    private var path: MutableList<Vector3f>? = null // Path to follow

    open fun Move(direction: Vector3f?) {
        print("hh")
    }

    fun getUUID(): UUID = id

    open fun move(terrain: Terrain?) {
        LuaEntityAPI.getEventsForType(this.name)
            ?.get("onMove")
            ?.let { it.call(CoerceJavaToLua.coerce(this)) }
    }


    open fun TakeDamage(damage: Float, Attacker: Mob?) {
        health -= damage

        LuaEntityAPI.getEventsForType(this.name)
            ?.get("onDamage")
            ?.let { it.call(CoerceJavaToLua.coerce(this), LuaValue.valueOf(damage.toInt())) }

        if (health <= 0f) Die()
    }

    open fun Die() {
        LuaEntityAPI.getEventsForType(this.name)
            ?.get("onDeath")
            ?.let { it.call(CoerceJavaToLua.coerce(this)) }
    }

    abstract fun attack(target: Mob?)

    protected fun findClosestEnemy(): Mob? {
        var closest: Mob? = null
        var closestDistance = Float.Companion.MAX_VALUE

        for (mob in getNearbyMobs(10.0f)) { // Implement this
            val distance: Float = position.distance(mob.position)
            if (distance < 2.0f && distance < closestDistance) {
                closest = mob
                closestDistance = distance
                log.info("Closes Mob FOUND! It's name is: {}", mob.name)
            }
        }
        return closest
    }

    private fun getNearbyMobs(detectionRadius: Float): MutableList<Mob> {
        val nearbyMobs: MutableList<Mob> = ArrayList<Mob>()

        for (entity in getEntities()) { // Assuming getEntities() returns all entities
            if (entity is Mob && entity !== this) { // Check if it's a Mob and not the player itself
                if (position.distance(entity.position) <= detectionRadius) {
                    nearbyMobs.add(entity)
                }
            }
        }
        return nearbyMobs
    }

    abstract fun updateAI(terrain: Terrain)

    fun setPath(newPath: List<Vector3f>?) {
        //println("SetPath to NewPath: $newPath")
        this.path = newPath?.toMutableList()
        //println("Path is: $path")
    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(Mob::class.java)
    }
}
